pkg/logs/diagnostic/message_receiver.go

```diff
@@ -0,0 +1,133 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package diagnostic
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/logs/message"
+)
+
+// MessageReceiver interface to handle messages for diagnostics
+type MessageReceiver interface {
+	HandleMessage(message.Message)
+}
+
+// BufferedMessageReceiver handles in coming log messages and makes them available for diagnostics
+type BufferedMessageReceiver struct {
+	inputChan chan message.Message
+	enabled   bool
+	m         sync.RWMutex
+}
+
+// Filters for processing log messages
+type Filters struct {
+	Name   string `json:"name"`
+	Type   string `json:"type"`
+	Source string `json:"source"`
+}
+
+// NewBufferedMessageReceiver creates a new MessageReceiver
+func NewBufferedMessageReceiver() *BufferedMessageReceiver {
+	return &BufferedMessageReceiver{
+		inputChan: make(chan message.Message, config.ChanSize),
+	}
+}
+
+// Start opens new input channel
+func (b *BufferedMessageReceiver) Start() {
+	b.inputChan = make(chan message.Message, config.ChanSize)
+}
+
+// Stop closes the input channel
+func (b *BufferedMessageReceiver) Stop() {
+	close(b.inputChan)
+}
+
+// Clear empties buffered messages
+func (b *BufferedMessageReceiver) clear() {
+	l := len(b.inputChan)
+	for i := 0; i < l; i++ {
+		<-b.inputChan
+	}
+}
+
+// SetEnabled start collecting log messages for diagnostics
+func (b *BufferedMessageReceiver) SetEnabled(e bool) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	if !e {
+		b.clear()
+	}
+	b.enabled = e
```

Comment:
Not sure it is an issue but `HandleMessage` can enqueue a message after line 65.